In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

print("Libraries imported successfully!")

Libraries imported successfully!


In [3]:
df=pd.read_csv("../data/raw/telco.csv")

PermissionError: [Errno 13] Permission denied: '../data/raw/telco.csv'

In [7]:
import os
print (os.listdir("../data/raw"))

['.gitkeep', 'telco.csv']


In [9]:
import os
print(os.path.isfile("../data/raw/telco.csv"))

False


In [10]:
print(os.isdir("../data/raw/telco.csv"))
print(os.path.isfile("../data/raw/telco.csv"))

AttributeError: module 'os' has no attribute 'isdir'

In [11]:
print(os.path.isdir("../data/raw/telco.csv"))
print(os.path.isfile("../data/raw/telco.csv"))

True
False


In [ ]:
import os
path = "../data/raw/telco.csv"

print("Exists:", os.path.exists(path))
print("is file:", os.path.isfile(path))
print("is directory:", os.path.isdir(path))

In [ ]:
import pandas as pd
df = pd.read_csv("../data/raw/telco.csv")

In [ ]:
print(os.pathisfile("../data/raw/telco.csv"))

In [ ]:
import os
print(os.path.isfile("../data/raw/telco.csv"))

In [ ]:
import os
print("Current working directory:")
print(os.getcwd())

In [ ]:
print("hello")

In [1]:
print("hello")

hello


In [2]:
import os
print(os.getcwd())

c:\Users\shubam\apexplanet-data-analytics\notebooks


In [3]:
print(os.path.exists("../data/raw/telco.csv"))

True


In [4]:
df = pd.read_csv("data/raw/telco.csv")

NameError: name 'pd' is not defined

In [5]:
import pandas as pd

In [6]:
df = pd.read_csv("../data/raw/telco.csv")

In [7]:
df.head()

,Customer ID,Gender,Age,Under 30,Senior Citizen,Married,Dependents,Number of Dependents,Country,State,...,Total Extra Data Charges,Total Long Distance Charges,Total Revenue,Satisfaction Score,Customer Status,Churn Label,Churn Score,CLTV,Churn Category,Churn Reason
0,8779-QRDMV,Male,78,No,Yes,No,No,0,United States,California,...,20,0.00,59.65,3,Churned,Yes,91,5433,Competitor,Competitor offered more data
1,7495-OOKFY,Female,74,No,Yes,Yes,Yes,1,United States,California,...,0,390.80,1024.10,3,Churned,Yes,69,5302,Competitor,Competitor made better offer
2,1658-BYGOY,Male,71,No,Yes,No,Yes,3,United States,California,...,0,203.94,1910.88,2,Churned,Yes,81,3179,Competitor,Competitor made better offer
3,4598-XLKNJ,Female,78,No,Yes,Yes,Yes,1,United States,California,...,0,494.00,2995.07,2,Churned,Yes,88,5337,Dissatisfaction,Limited range of services
4,4846-WHAFZ,Female,80,No,Yes,Yes,Yes,1,United States,California,...,0,234.21,3102.36,2,Churned,Yes,67,2793,Price,Extra data charges


In [8]:
df.shape

(7043, 50)

In [9]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 50 columns):
 #   Column                             Non-Null Count  Dtype  
---  ------                             --------------  -----  
 0   Customer ID                        7043 non-null   object 
 1   Gender                             7043 non-null   object 
 2   Age                                7043 non-null   int64  
 3   Under 30                           7043 non-null   object 
 4   Senior Citizen                     7043 non-null   object 
 5   Married                            7043 non-null   object 
 6   Dependents                         7043 non-null   object 
 7   Number of Dependents               7043 non-null   int64  
 8   Country                            7043 non-null   object 
 9   State                              7043 non-null   object 
 10  City                               7043 non-null   object 
 11  Zip Code                           7043 non-null   int64

In [10]:
import pandas as pd
df = pd.read_csv("../data/raw/telco.csv")
df.head()

,Customer ID,Gender,Age,Under 30,Senior Citizen,Married,Dependents,Number of Dependents,Country,State,...,Total Extra Data Charges,Total Long Distance Charges,Total Revenue,Satisfaction Score,Customer Status,Churn Label,Churn Score,CLTV,Churn Category,Churn Reason
0,8779-QRDMV,Male,78,No,Yes,No,No,0,United States,California,...,20,0.00,59.65,3,Churned,Yes,91,5433,Competitor,Competitor offered more data
1,7495-OOKFY,Female,74,No,Yes,Yes,Yes,1,United States,California,...,0,390.80,1024.10,3,Churned,Yes,69,5302,Competitor,Competitor made better offer
2,1658-BYGOY,Male,71,No,Yes,No,Yes,3,United States,California,...,0,203.94,1910.88,2,Churned,Yes,81,3179,Competitor,Competitor made better offer
3,4598-XLKNJ,Female,78,No,Yes,Yes,Yes,1,United States,California,...,0,494.00,2995.07,2,Churned,Yes,88,5337,Dissatisfaction,Limited range of services
4,4846-WHAFZ,Female,80,No,Yes,Yes,Yes,1,United States,California,...,0,234.21,3102.36,2,Churned,Yes,67,2793,Price,Extra data charges


In [ ]:
df.shape

In [ ]:
import pandas as pd

In [ ]:
df = pd.read_csv("../data/raw/telco.csv")

In [1]:
import pandas as pd

df = pd.read_csv("../data/raw/telco.csv")

In [2]:
df.head

<bound method NDFrame.head of      Customer ID  Gender  Age Under 30 Senior Citizen Married Dependents  \
0     8779-QRDMV    Male   78       No            Yes      No         No   
1     7495-OOKFY  Female   74       No            Yes     Yes        Yes   
2     1658-BYGOY    Male   71       No            Yes      No        Yes   
3     4598-XLKNJ  Female   78       No            Yes     Yes        Yes   
4     4846-WHAFZ  Female   80       No            Yes     Yes        Yes   
...          ...     ...  ...      ...            ...     ...        ...   
7038  2569-WGERO  Female   30       No             No      No         No   
7039  6840-RESVB    Male   38       No             No     Yes        Yes   
7040  2234-XADUH  Female   30       No             No     Yes        Yes   
7041  4801-JZAZL  Female   32       No             No     Yes        Yes   
7042  3186-AJIEK    Male   44       No             No      No         No   

      Number of Dependents        Country       State  ..

In [3]:
df.shape

(7043, 50)

In [4]:
df.columns

Index(['Customer ID', 'Gender', 'Age', 'Under 30', 'Senior Citizen', 'Married',
       'Dependents', 'Number of Dependents', 'Country', 'State', 'City',
       'Zip Code', 'Latitude', 'Longitude', 'Population', 'Quarter',
       'Referred a Friend', 'Number of Referrals', 'Tenure in Months', 'Offer',
       'Phone Service', 'Avg Monthly Long Distance Charges', 'Multiple Lines',
       'Internet Service', 'Internet Type', 'Avg Monthly GB Download',
       'Online Security', 'Online Backup', 'Device Protection Plan',
       'Premium Tech Support', 'Streaming TV', 'Streaming Movies',
       'Streaming Music', 'Unlimited Data', 'Contract', 'Paperless Billing',
       'Payment Method', 'Monthly Charge', 'Total Charges', 'Total Refunds',
       'Total Extra Data Charges', 'Total Long Distance Charges',
       'Total Revenue', 'Satisfaction Score', 'Customer Status', 'Churn Label',
       'Churn Score', 'CLTV', 'Churn Category', 'Churn Reason'],
      dtype='object')

In [5]:
for col in df.columns:
    print(col)

Customer ID
Gender
Age
Under 30
Senior Citizen
Married
Dependents
Number of Dependents
Country
State
City
Zip Code
Latitude
Longitude
Population
Quarter
Referred a Friend
Number of Referrals
Tenure in Months
Offer
Phone Service
Avg Monthly Long Distance Charges
Multiple Lines
Internet Service
Internet Type
Avg Monthly GB Download
Online Security
Online Backup
Device Protection Plan
Premium Tech Support
Streaming TV
Streaming Movies
Streaming Music
Unlimited Data
Contract
Paperless Billing
Payment Method
Monthly Charge
Total Charges
Total Refunds
Total Extra Data Charges
Total Long Distance Charges
Total Revenue
Satisfaction Score
Customer Status
Churn Label
Churn Score
CLTV
Churn Category
Churn Reason


In [6]:
df.info

<bound method DataFrame.info of      Customer ID  Gender  Age Under 30 Senior Citizen Married Dependents  \
0     8779-QRDMV    Male   78       No            Yes      No         No   
1     7495-OOKFY  Female   74       No            Yes     Yes        Yes   
2     1658-BYGOY    Male   71       No            Yes      No        Yes   
3     4598-XLKNJ  Female   78       No            Yes     Yes        Yes   
4     4846-WHAFZ  Female   80       No            Yes     Yes        Yes   
...          ...     ...  ...      ...            ...     ...        ...   
7038  2569-WGERO  Female   30       No             No      No         No   
7039  6840-RESVB    Male   38       No             No     Yes        Yes   
7040  2234-XADUH  Female   30       No             No     Yes        Yes   
7041  4801-JZAZL  Female   32       No             No     Yes        Yes   
7042  3186-AJIEK    Male   44       No             No      No         No   

      Number of Dependents        Country       State  

In [7]:
df.describe

<bound method NDFrame.describe of      Customer ID  Gender  Age Under 30 Senior Citizen Married Dependents  \
0     8779-QRDMV    Male   78       No            Yes      No         No   
1     7495-OOKFY  Female   74       No            Yes     Yes        Yes   
2     1658-BYGOY    Male   71       No            Yes      No        Yes   
3     4598-XLKNJ  Female   78       No            Yes     Yes        Yes   
4     4846-WHAFZ  Female   80       No            Yes     Yes        Yes   
...          ...     ...  ...      ...            ...     ...        ...   
7038  2569-WGERO  Female   30       No             No      No         No   
7039  6840-RESVB    Male   38       No             No     Yes        Yes   
7040  2234-XADUH  Female   30       No             No     Yes        Yes   
7041  4801-JZAZL  Female   32       No             No     Yes        Yes   
7042  3186-AJIEK    Male   44       No             No      No         No   

      Number of Dependents        Country       State

In [8]:
df.describe(include="object")

,Customer ID,Gender,Under 30,Senior Citizen,Married,Dependents,Country,State,City,Quarter,...,Streaming Movies,Streaming Music,Unlimited Data,Contract,Paperless Billing,Payment Method,Customer Status,Churn Label,Churn Category,Churn Reason
count,7043,7043,7043,7043,7043,7043,7043,7043,7043,7043,...,7043,7043,7043,7043,7043,7043,7043,7043,1869,1869
unique,7043,2,2,2,2,2,1,1,1106,1,...,2,2,2,3,2,3,3,2,5,20
top,3186-AJIEK,Male,No,No,No,No,United States,California,Los Angeles,Q3,...,No,No,Yes,Month-to-Month,Yes,Bank Withdrawal,Stayed,No,Competitor,Competitor had better devices
freq,1,3555,5642,5901,3641,5416,7043,7043,293,7043,...,4311,4555,4745,3610,4171,3909,4720,5174,841,313


In [9]:
df.isnull().sum()

Customer ID                             0
Gender                                  0
Age                                     0
Under 30                                0
Senior Citizen                          0
Married                                 0
Dependents                              0
Number of Dependents                    0
Country                                 0
State                                   0
City                                    0
Zip Code                                0
Latitude                                0
Longitude                               0
Population                              0
Quarter                                 0
Referred a Friend                       0
Number of Referrals                     0
Tenure in Months                        0
Offer                                3877
Phone Service                           0
Avg Monthly Long Distance Charges       0
Multiple Lines                          0
Internet Service                  

In [10]:
df.isnull().sum()[df.insull().sum()>0]

AttributeError: 'DataFrame' object has no attribute 'insull'

In [11]:
df.isnull().sum()[df.isnull.sum()>0]

AttributeError: 'function' object has no attribute 'sum'

In [12]:
df.isnull().sum()[df.isnull().sum()>0]

Offer             3877
Internet Type     1526
Churn Category    5174
Churn Reason      5174
dtype: int64

In [13]:
df.duplicated().sum()

np.int64(0)

In [14]:
missing =df.isnull().sum()
missing[missing >0]

Offer             3877
Internet Type     1526
Churn Category    5174
Churn Reason      5174
dtype: int64

In [15]:
df.duplicated().sum()

np.int64(0)

# Dataset Information

## Dataset Name
IBM Telco Customer Churn Dataset

## Source
Kaggle

## Description
This dataset contains customer information from a telecommunications company. It includes demographic details, services subscribed, billing information, customer satisfaction, and whether the customer churned.

## Rows
7043

## Columns
50

## Target Variable
Churn Label

## Missing Values
- Offer
- Internet Type
- Churn Category
- Churn Reason

## Duplicate Rows
0

In [16]:
df_clean = df.copy()

In [17]:
df_clean.isnull().sum()

Customer ID                             0
Gender                                  0
Age                                     0
Under 30                                0
Senior Citizen                          0
Married                                 0
Dependents                              0
Number of Dependents                    0
Country                                 0
State                                   0
City                                    0
Zip Code                                0
Latitude                                0
Longitude                               0
Population                              0
Quarter                                 0
Referred a Friend                       0
Number of Referrals                     0
Tenure in Months                        0
Offer                                3877
Phone Service                           0
Avg Monthly Long Distance Charges       0
Multiple Lines                          0
Internet Service                  

In [18]:
df_clean["offer"] =df_clean["offer"].fillna("No Offer")

df_clean["Internet Type"] = df_clean["Internet Type"].fillna("No Internet")

df_clean["Churn category"] = df_clean["Churn Category"].fillna("Not Churned")

df_clean["Churn Reason"] = df_clean["Churn Reason"].fillna("Not Applicable")

KeyError: 'offer'

In [19]:
df_clean["Offer"] =df_clean["offer"].fillna("No Offer")

df_clean["Internet Type"] = df_clean["Internet Type"].fillna("No Internet")

df_clean["Churn category"] = df_clean["Churn Category"].fillna("Not Churned")

df_clean["Churn Reason"] = df_clean["Churn Reason"].fillna("Not Applicable")

KeyError: 'offer'

In [20]:
df_clean["Offer"] =df_clean["Offer"].fillna("No Offer")

df_clean["Internet Type"] = df_clean["Internet Type"].fillna("No Internet")

df_clean["Churn category"] = df_clean["Churn Category"].fillna("Not Churned")

df_clean["Churn Reason"] = df_clean["Churn Reason"].fillna("Not Applicable")

In [21]:
df_clean.isnull().sum()

Customer ID                             0
Gender                                  0
Age                                     0
Under 30                                0
Senior Citizen                          0
Married                                 0
Dependents                              0
Number of Dependents                    0
Country                                 0
State                                   0
City                                    0
Zip Code                                0
Latitude                                0
Longitude                               0
Population                              0
Quarter                                 0
Referred a Friend                       0
Number of Referrals                     0
Tenure in Months                        0
Offer                                   0
Phone Service                           0
Avg Monthly Long Distance Charges       0
Multiple Lines                          0
Internet Service                  

In [22]:
df_clean["Churn Category"] = df_clean["Churn Category"].fillna("Not Churned")

In [23]:
df_clean.isnull().sum()

Customer ID                          0
Gender                               0
Age                                  0
Under 30                             0
Senior Citizen                       0
Married                              0
Dependents                           0
Number of Dependents                 0
Country                              0
State                                0
City                                 0
Zip Code                             0
Latitude                             0
Longitude                            0
Population                           0
Quarter                              0
Referred a Friend                    0
Number of Referrals                  0
Tenure in Months                     0
Offer                                0
Phone Service                        0
Avg Monthly Long Distance Charges    0
Multiple Lines                       0
Internet Service                     0
Internet Type                        0
Avg Monthly GB Download  

In [24]:
df_clean.duplicated().sum()

np.int64(0)

In [25]:
df_clean.columns = (
    df_clean.columns
    .str.strip()
    .str.lower()
    .str.replace(" ","_")
)

In [26]:
df_clean.columns

Index(['customer_id', 'gender', 'age', 'under_30', 'senior_citizen', 'married',
       'dependents', 'number_of_dependents', 'country', 'state', 'city',
       'zip_code', 'latitude', 'longitude', 'population', 'quarter',
       'referred_a_friend', 'number_of_referrals', 'tenure_in_months', 'offer',
       'phone_service', 'avg_monthly_long_distance_charges', 'multiple_lines',
       'internet_service', 'internet_type', 'avg_monthly_gb_download',
       'online_security', 'online_backup', 'device_protection_plan',
       'premium_tech_support', 'streaming_tv', 'streaming_movies',
       'streaming_music', 'unlimited_data', 'contract', 'paperless_billing',
       'payment_method', 'monthly_charge', 'total_charges', 'total_refunds',
       'total_extra_data_charges', 'total_long_distance_charges',
       'total_revenue', 'satisfaction_score', 'customer_status', 'churn_label',
       'churn_score', 'cltv', 'churn_category', 'churn_reason',
       'churn_category'],
      dtype='object')

In [1]:
df.head()

NameError: name 'df' is not defined

In [2]:
df_clean.head()

NameError: name 'df_clean' is not defined

In [3]:
import pandas as pd

df = pd.read_csv("../data/raw/telco.csv")

In [4]:
df_clean.head()

NameError: name 'df_clean' is not defined

In [5]:
df_clean = df.copy()

In [6]:
df_clean.head()

,Customer ID,Gender,Age,Under 30,Senior Citizen,Married,Dependents,Number of Dependents,Country,State,...,Total Extra Data Charges,Total Long Distance Charges,Total Revenue,Satisfaction Score,Customer Status,Churn Label,Churn Score,CLTV,Churn Category,Churn Reason
0,8779-QRDMV,Male,78,No,Yes,No,No,0,United States,California,...,20,0.00,59.65,3,Churned,Yes,91,5433,Competitor,Competitor offered more data
1,7495-OOKFY,Female,74,No,Yes,Yes,Yes,1,United States,California,...,0,390.80,1024.10,3,Churned,Yes,69,5302,Competitor,Competitor made better offer
2,1658-BYGOY,Male,71,No,Yes,No,Yes,3,United States,California,...,0,203.94,1910.88,2,Churned,Yes,81,3179,Competitor,Competitor made better offer
3,4598-XLKNJ,Female,78,No,Yes,Yes,Yes,1,United States,California,...,0,494.00,2995.07,2,Churned,Yes,88,5337,Dissatisfaction,Limited range of services
4,4846-WHAFZ,Female,80,No,Yes,Yes,Yes,1,United States,California,...,0,234.21,3102.36,2,Churned,Yes,67,2793,Price,Extra data charges
